internal/impl/pure/rate_limit_local.go

```diff
@@ -39,55 +43,116 @@ func newLocalRatelimitFromConfig(conf *service.ParsedConfig) (*localRatelimit, e
 	if err != nil {
 		return nil, err
 	}
+	byteSize, err := conf.FieldInt("byte_size")
+	if err != nil {
+		return nil, err
+	}
 	interval, err := conf.FieldDuration("interval")
 	if err != nil {
 		return nil, err
 	}
-	return newLocalRatelimit(count, interval)
+	return newLocalRatelimit(count, byteSize, interval)
 }
 
 //------------------------------------------------------------------------------
 
 type localRatelimit struct {
-	mut         sync.Mutex
-	bucket      int
-	lastRefresh time.Time
+	mut           sync.Mutex // TODO: We should rather be using atomics as opposed to locking
+	bucket        int
+	byteBucket    int
+	exceededLimit bool
+	lastRefresh   time.Time
 
-	size   int
-	period time.Duration
+	size     int
+	byteSize int
+	period   time.Duration
 }
 
-func newLocalRatelimit(count int, interval time.Duration) (*localRatelimit, error) {
-	if count <= 0 {
-		return nil, errors.New("count must be larger than zero")
+func newLocalRatelimit(count, byteSize int, interval time.Duration) (*localRatelimit, error) {
+	if byteSize < 0 || count < 0 {
+		return nil, errors.New("neither byte size nor count can be negative")
+	}
+
+	if byteSize == 0 && count == 0 {
+		return nil, errors.New("either count or byte size must be larger than zero")
 	}
+
 	return &localRatelimit{
-		bucket:      count,
-		lastRefresh: time.Now(),
-		size:        count,
-		period:      interval,
+		bucket:        count,
+		byteBucket:    byteSize,
+		exceededLimit: false,
+		lastRefresh:   time.Now(),
+
+		size:     count,
+		byteSize: byteSize,
+		period:   interval,
 	}, nil
 }
 
 func (r *localRatelimit) Access(ctx context.Context) (time.Duration, error) {
 	r.mut.Lock()
-	r.bucket--
+	defer r.mut.Unlock()
+
+	// Rate limiting count is enabled
+	if r.size > 0 {
+		r.bucket--
+	}
 
 	if r.bucket < 0 {
-		r.bucket = 0
-		remaining := r.period - time.Since(r.lastRefresh)
+		r.exceededLimit = true
+	}
 
-		if remaining > 0 {
-			r.mut.Unlock()
-			return remaining, nil
-		}
-		r.bucket = r.size - 1
-		r.lastRefresh = time.Now()
+	if !r.exceededLimit {
+		return 0, nil
 	}
-	r.mut.Unlock()
+
+	remaining := r.period - time.Since(r.lastRefresh)
+	if remaining > 0 {
+		return remaining, nil
+	}
+
+	// The interval has passed so reset
+	r.refresh()
 	return 0, nil
 }
 
+// Add increments the observation or message-bytes counter. Returns true if this
+// triggers the conditions of the rate-limiter.
+func (r *localRatelimit) Add(ctx context.Context, parts ...*message.Part) bool {
+	r.mut.Lock()
+	defer r.mut.Unlock()
+
+	if time.Since(r.lastRefresh) >= r.period {
+		r.refresh()
+	}
+
+	if r.exceededLimit {
+		return true
+	}
+
+	// Rate limiting bytes is enabled
+	if r.byteSize > 0 && len(parts) != 0 {
```

Comment:
```suggestion
	if r.byteSize > 0 {
```

nit: is that `len(parts) != 0` required if it is then subsequently in the for loop condition?

Reply:
Yeah this isn't necesary. Will change